File: tqec_optimizer/relocation/module_factory.py
from .module import Module

from ..vector3d import Vector3D
from ..node import Node, Joint
from ..edge import Edge, CrossEdge


class ModuleFactory:
    def __init__(self, type_, loop, injector_list):
        self._type = type_
        self._loop = loop
        self._injector_list = injector_list
        self._module = Module(self._loop.id)

    def create(self):
        self.__create_frame()
        self.__create_cross_edge()
        self.__create_injector()
        self._module.update()

        return self._module

    def __create_frame(self):
        has_injector = True if len(self._injector_list) > 0 else False
        cross_edge_num = max(1, len(self._loop.cross_list))
        injector_num = max(0, len(self._loop.injector_list) - 1)
        node_array = []
        pos = Vector3D(0, 0, 0)

        # create frame node
        node_array.append(self.__new_node("frame", pos))
        pos.incy(2)
        node_array.append(self.__new_node("frame", pos))
        for n in range(0, cross_edge_num + injector_num):
            pos.incz(2)
            node_array.append(self.__new_node("frame", pos))
        pos.decy(2)
        node_array.append(self.__new_node("frame", pos))
        for n in range(0, cross_edge_num + injector_num - 1):
            pos.decz(2)
            node_array.append(self.__new_node("frame", pos))

        # create frame edge
        last_node = None
        for n, node in enumerate(node_array):
            if last_node is not None:
                category = self._injector_list.pop() \
                    if 2 + cross_edge_num + injector_num == n and has_injector \
                    else "edge"
                edge = self.__new__edge("frame", node, last_node, category, self._loop.id)
                self._module.add_frame_edge(edge)
            last_node = node
        edge = self.__new__edge("frame", node_array[0], last_node, "edge", self._loop.id)
        self._module.add_frame_edge(edge)

    def __create_cross_edge(self):
        type_ = "dual" if self._type == "primal" else "primal"
        pos1, pos2 = Vector3D(1, 1, 1), Vector3D(-1, 1, 1)
        for cross_edge_id in self._loop.cross_list:
            joint1 = self.__new_node("cross", pos1, type_, cross_edge_id, set(self._loop.cross_list))
            joint2 = self.__new_node("cross", pos2, type_, cross_edge_id, set(self._loop.cross_list))
            cross_edge = self.__new__edge("cross", joint1, joint2, "edge", cross_edge_id, self._loop.id)
            self._module.add_cross_edge(cross_edge)
            self._module.add_cross_id(cross_edge_id)
            self._module.add_joint_pair((joint1, joint2, cross_edge))
            pos1.incz(2)
            pos2.incz(2)

    def __create_injector(self):
        self._injector_list.clear()
        cross_edge_num = max(1, len(self._loop.cross_list))
        injector_num = max(0, len(self._loop.injector_list) - 1)
        max_z = 2 * (cross_edge_num + injector_num)
        for n in range(1, injector_num + 1):
            node1 = self.__frame_node(0, 0, max_z - 2 * n)
            node2 = self.__frame_node(0, 2, max_z - 2 * n)
            injector = self.__new__edge("frame", node1, node2, "pin", self._loop.id)
            self._module.add_frame_edge(injector)

    def __frame_node(self, x, y, z):
        for node in self._module.frame_node_list:
            if node.x == x and node.y == y and node.z == z:
                return node
        return None

    def __new_node(self, category, pos, type_=None, id_=0, id_set=None):
        if category == "frame":
            node = Node(pos.x, pos.y, pos.z, self._loop.id, self._type)
            self._module.add_frame_node(node)
        else:
            node = Joint(pos.x, pos.y, pos.z, id_, type_)
            node.set_id_set(id_set)
            self._module.add_cross_node(node)

        return node

    @staticmethod
    def __new__edge(class_, node1, node2, category, id_=0, module_id=0):
        if class_ == "frame":
            edge = Edge(node1, node2, category, id_)
            node1.add_edge(edge)
            node2.add_edge(edge)
        else:
            edge = CrossEdge(node1, node2, category, id_, module_id)
            node1.add_edge(edge)
            node2.add_edge(edge)

        return edge

File: src/topology/src/module/module_factory.cpp
#include "module_factory.hpp"

Module ModuleFactory::Create(const Loop& loop) {
    std::vector<NodePtr> frame_nodes, cross_nodes;
    std::vector<EdgePtr> frame_edges, cross_edges;

    CreateFrame(loop, frame_nodes, frame_edges);
    CreateCross(loop, cross_nodes, cross_edges);

    return Module(loop.id(), loop.cross_list(),
                  frame_nodes, cross_nodes,
                  frame_edges, cross_edges);
}

void ModuleFactory::CreateFrame(const Loop& loop,
                                std::vector<NodePtr>& frame_nodes,
                                std::vector<EdgePtr>& frame_edges) {
    const int id = loop.id();
    const std::string type = loop.type();
    const int cross_edge_num = std::max(1, static_cast<int>(loop.cross_list().size()));
    const int injector_num = loop.pin_num() + loop.cap_num();
    const int frame_length = cross_edge_num + std::max(0, injector_num - 1);

    NodePtr upper_node1, upper_node2, lower_node1, lower_node2;
    Vector3D upper_pos = Vector3D(0, 2, 0);
    Vector3D lower_pos = Vector3D(0, 0, 0);

    // create frame
    frame_nodes.push_back(upper_node1 = std::make_shared<Node>(id, upper_pos, type));
    frame_nodes.push_back(lower_node1 = std::make_shared<Node>(id, lower_pos, type));
    frame_edges.push_back(std::make_shared<Edge>(id, "edge", upper_node1, lower_node1));

    for (int n = 0; n < frame_length; ++n) {
        upper_pos.z += 2;
        lower_pos.z += 2;
        upper_node2 = upper_node1;
        lower_node2 = lower_node1;
        frame_nodes.push_back(upper_node1 = std::make_shared<Node>(id, upper_pos, type));
        frame_nodes.push_back(lower_node1 = std::make_shared<Node>(id, lower_pos, type));
        frame_edges.push_back(std::make_shared<Edge>(id, "edge", upper_node1, upper_node2));
        frame_edges.push_back(std::make_shared<Edge>(id, "edge", lower_node1, lower_node2));
        if (injector_num > 0 && n >= cross_edge_num - 1) {
            frame_edges.push_back(std::make_shared<Edge>(id, "injector", upper_node1, lower_node1));
        }
    }

    return;
}

void ModuleFactory::CreateCross(const Loop& loop,
                                std::vector<NodePtr>& cross_nodes,
                                std::vector<EdgePtr>& cross_edges) {
    const std::string type = (loop.type() == "primal") ? "dual" : "primal";
    Vector3D pos1 = Vector3D(1, 1, 1);
    Vector3D pos2 = Vector3D(-1, 1, 1);
    NodePtr joint1, joint2;
    for (const int cross_id : loop.cross_list()) {
        cross_nodes.push_back(joint1 = std::make_shared<Node>(cross_id, pos1, type));
        cross_nodes.push_back(joint2 = std::make_shared<Node>(cross_id, pos2, type));
        cross_edges.push_back(std::make_shared<Edge>(cross_id, "edge", joint1, joint2));
        pos1.add(0, 0, 1);
        pos2.add(0, 0, 1);
    }

    return;
}